Use latest version of sockjs-client

harryadel opened the pull request:
This seems to be a [long requested](https://github.com/meteor/meteor-feature-requests/issues/216) fix in Meteor, dicussed in https://github.com/meteor/meteor/pull/9215, https://github.com/meteor/meteor/issues/4114, https://github.com/meteor/meteor/pull/9985

With initial go ahead from the Meteor core team which was only held back by the need to further investigation the addition of JSON3 and jump in bundle size.  
https://github.com/meteor/meteor/issues/4114#issuecomment-273472853
>I guess I should clarify that the size increase stated was in development mode, so less alarming as the production bundle would be minified. Still a big jump. I just did a quick test of the (minified) production size and it's only an increase of 10KB on the final JS resource bundle, so a lot less alarming.
>
>I'll still look into it though as it's not really clear at first glance or from the Changelog. The file size difference of the uncompressed distribution in their repo is 282KB between 0.3.4 and 1.1.1 – a huge increase. JSON2 was commented out of the old version on the Meteor side but that only affects the minified production version for the sake of this comparison. The new version comes with JSON3.

>Anyhow, more on this soon.

It's a humble attempt and I'm open to feedback, thanks!

StorytellerCZ commented on packages/socket-stream-client/sockjs.js:
```diff
@@ -0,0 +1 @@
+import SockJS from "sockjs-client";
```
Shouldn't this be an export?

StorytellerCZ commented on packages/socket-stream-client/browser.js:
```diff
@@ -8,7 +8,7 @@ import { StreamClientCommon } from "./common.js";
 // below (in favor of SockJS), but will ensure maximum compatibility for
 // clients stuck in unusual networking environments.
-import "./sockjs-0.3.4.js";
+import "./sockjs";
```
You could just do `import SockJS from "sockjs-client";`.

harryadel commented:
@StorytellerCZ no matter how `sockjs-client` is imported/exported, it always resulted in an error in one of the self tests `add packages client archs` which you can run locally using:
```
./meteor self-test "add packages client archs"
``` 
Copying the contents of the CDN `https://cdn.jsdelivr.net/npm/sockjs-client@1/dist/sockjs.min.js` into `sockjs.js` seems to fix the problem. What do you think?

StorytellerCZ reviewed (approved):
Shouldn't we now just use the SockJS version from NPM?

harryadel commented:
@StorytellerCZ yeah, using the NPM module would be a lot more elegant but I couldn't get it working as due to the reasons illustrated above. If you can get it working, I'd be happy to close this PR.

StorytellerCZ commented:
@harryadel Getting the new version working for now is good enough. Other improvements can come later. We don't have to have everything at once. I'm for the incremental approach.

harryadel commented:
Here's what you get by running `git diff` on the current [sockjs-0.3.4](https://github.com/meteor/meteor/blob/devel/packages/socket-stream-client/sockjs-0.3.4.js) replacing the contents with the [official 0.3.4](https://github.com/sockjs/sockjs-client/releases/tag/v0.3.4).

```diff
diff --git a/packages/socket-stream-client/sockjs-0.3.4.js b/packages/socket-stream-client/sockjs-0.3.4.js
index f4bc6ac39..585215cb5 100644
--- a/packages/socket-stream-client/sockjs-0.3.4.js
+++ b/packages/socket-stream-client/sockjs-0.3.4.js
@@ -1,5 +1,3 @@
-// XXX METEOR changes in <METEOR>
-
 /* SockJS client, version 0.3.4, http://sockjs.org, MIT License
 
 Copyright (c) 2011-2012 VMware, Inc.

@@ -23,10 +21,9 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */
 
-// <METEOR> Commented out JSO implementation (use json package instead).
 // JSON2 by Douglas Crockford (minified).
-// var JSON;JSON||(JSON={}),function(){function str(a,b){var c,d,e,f,g=gap,h,i=b[a];i&&typeof i=="object"&&typeof i.toJSON=="function"&&(i=i.toJSON(a)),typeof rep=="fu
nction"&&(i=rep.call(b,a,i));switch(typeof i){case"string":return quote(i);case"number":return isFinite(i)?String(i):"null";case"boolean":case"null":return String(i);ca
se"object":if(!i)return"null";gap+=indent,h=[];if(Object.prototype.toString.apply(i)==="[object Array]"){f=i.length;for(c=0;c<f;c+=1)h[c]=str(c,i)||"null";e=h.length===
0?"[]":gap?"[\n"+gap+h.join(",\n"+gap)+"\n"+g+"]":"["+h.join(",")+"]",gap=g;return e}if(rep&&typeof rep=="object"){f=rep.length;for(c=0;c<f;c+=1)typeof rep[c]=="string"
&&(d=rep[c],e=str(d,i),e&&h.push(quote(d)+(gap?": ":":")+e))}else for(d in i)Object.prototype.hasOwnProperty.call(i,d)&&(e=str(d,i),e&&h.push(quote(d)+(gap?": ":":")+e)
);e=h.length===0?"{}":gap?"{\n"+gap+h.join(",\n"+gap)+"\n"+g+"}":"{"+h.join(",")+"}",gap=g;return e}}function quote(a){escapable.lastIndex=0;return escapable.test(a)?'"
'+a.replace(escapable,function(a){var b=meta[a];return typeof b=="string"?b:"\\u"+("0000"+a.charCodeAt(0).toString(16)).slice(-4)})+'"':'"'+a+'"'}function f(a){return a
<10?"0"+a:a}"use strict",typeof Date.prototype.toJSON!="function"&&(Date.prototype.toJSON=function(a){return isFinite(this.valueOf())?this.getUTCFullYear()+"-"+f(this.g
etUTCMonth()+1)+"-"+f(this.getUTCDate())+"T"+f(this.getUTCHours())+":"+f(this.getUTCMinutes())+":"+f(this.getUTCSeconds())+"Z":null},String.prototype.toJSON=Number.prot
otype.toJSON=Boolean.prototype.toJSON=function(a){return this.valueOf()});var cx=/[\u0000\u00ad\u0600-\u0604\u070f\u17b4\u17b5\u200c-\u200f\u2028-\u202f\u2060-\u206f\uf
eff\ufff0-\uffff]/g,escapable=/[\\\"\x00-\x1f\x7f-\x9f\u00ad\u0600-\u0604\u070f\u17b4\u17b5\u200c-\u200f\u2028-\u202f\u2060-\u206f\ufeff\ufff0-\uffff]/g,gap,indent,meta
={"\b":"\\b","\t":"\\t","\n":"\\n","\f":"\\f","\r":"\\r",'"':'\\"',"\\":"\\\\"},rep;typeof JSON.stringify!="function"&&(JSON.stringify=function(a,b,c){var d;gap="",inde
nt="";if(typeof c=="number")for(d=0;d<c;d+=1)indent+=" ";else typeof c=="string"&&(indent=c);rep=b;if(!b||typeof b=="function"||typeof b=="object"&&typeof b.length=="nu
mber")return str("",{"":a});throw new Error("JSON.stringify")}),typeof JSON.parse!="function"&&(JSON.parse=function(text,reviver){function walk(a,b){var c,d,e=a[b];if(e
&&typeof e=="object")for(c in e)Object.prototype.hasOwnProperty.call(e,c)&&(d=walk(e,c),d!==undefined?e[c]=d:delete e[c]);return reviver.call(a,b,e)}var j;text=String(t
ext),cx.lastIndex=0,cx.test(text)&&(text=text.replace(cx,function(a){return"\\u"+("0000"+a.charCodeAt(0).toString(16)).slice(-4)}));if(/^[\],:{}\s]*$/.test(text.replace
(/\\(?:["\\\/bfnrt]|u[0-9a-fA-F]{4})/g,"@").replace(/"[^"\\\n\r]*"|true|false|null|-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?/g,"]").replace(/(?:^|:|,)(?:\s*\[)+/g,""))){j=eval(
"("+text+")");return typeof reviver=="function"?walk({"":j},""):j}throw new SyntaxError("JSON.parse")})}()
-// </METEOR>

+var JSON;JSON||(JSON={}),function(){function str(a,b){var c,d,e,f,g=gap,h,i=b[a];i&&typeof i=="object"&&typeof i.toJSON=="function"&&(i=i.toJSON(a)),typeof rep=="funct
ion"&&(i=rep.call(b,a,i));switch(typeof i){case"string":return quote(i);case"number":return isFinite(i)?String(i):"null";case"boolean":case"null":return String(i);case"
object":if(!i)return"null";gap+=indent,h=[];if(Object.prototype.toString.apply(i)==="[object Array]"){f=i.length;for(c=0;c<f;c+=1)h[c]=str(c,i)||"null";e=h.length===0?"
[]":gap?"[\n"+gap+h.join(",\n"+gap)+"\n"+g+"]":"["+h.join(",")+"]",gap=g;return e}if(rep&&typeof rep=="object"){f=rep.length;for(c=0;c<f;c+=1)typeof rep[c]=="string"&&(
d=rep[c],e=str(d,i),e&&h.push(quote(d)+(gap?": ":":")+e))}else for(d in i)Object.prototype.hasOwnProperty.call(i,d)&&(e=str(d,i),e&&h.push(quote(d)+(gap?": ":":")+e));e
=h.length===0?"{}":gap?"{\n"+gap+h.join(",\n"+gap)+"\n"+g+"}":"{"+h.join(",")+"}",gap=g;return e}}function quote(a){escapable.lastIndex=0;return escapable.test(a)?'"'+a
.replace(escapable,function(a){var b=meta[a];return typeof b=="string"?b:"\\u"+("0000"+a.charCodeAt(0).toString(16)).slice(-4)})+'"':'"'+a+'"'}function f(a){return a<10
?"0"+a:a}"use strict",typeof Date.prototype.toJSON!="function"&&(Date.prototype.toJSON=function(a){return isFinite(this.valueOf())?this.getUTCFullYear()+"-"+f(this.getU
TCMonth()+1)+"-"+f(this.getUTCDate())+"T"+f(this.getUTCHours())+":"+f(this.getUTCMinutes())+":"+f(this.getUTCSeconds())+"Z":null},String.prototype.toJSON=Number.prototy
pe.toJSON=Boolean.prototype.toJSON=function(a){return this.valueOf()});var cx=/[\u0000\u00ad\u0600-\u0604\u070f\u17b4\u17b5\u200c-\u200f\u2028-\u202f\u2060-\u206f\ufeff
\ufff0-\uffff]/g,escapable=/[\\\"\x00-\x1f\x7f-\x9f\u00ad\u0600-\u0604\u070f\u17b4\u17b5\u200c-\u200f\u2028-\u202f\u2060-\u206f\ufeff\ufff0-\uffff]/g,gap,indent,meta={"
\b":"\\b","\t":"\\t","\n":"\\n","\f":"\\f","\r":"\\r",'"':'\\"',"\\":"\\\\"},rep;typeof JSON.stringify!="function"&&(JSON.stringify=function(a,b,c){var d;gap="",indent=
"";if(typeof c=="number")for(d=0;d<c;d+=1)indent+=" ";else typeof c=="string"&&(indent=c);rep=b;if(!b||typeof b=="function"||typeof b=="object"&&typeof b.length=="numbe
r")return str("",{"":a});throw new Error("JSON.stringify")}),typeof JSON.parse!="function"&&(JSON.parse=function(text,reviver){function walk(a,b){var c,d,e=a[b];if(e&&t
ypeof e=="object")for(c in e)Object.prototype.hasOwnProperty.call(e,c)&&(d=walk(e,c),d!==undefined?e[c]=d:delete e[c]);return reviver.call(a,b,e)}var j;text=String(text
),cx.lastIndex=0,cx.test(text)&&(text=text.replace(cx,function(a){return"\\u"+("0000"+a.charCodeAt(0).toString(16)).slice(-4)}));if(/^[\],:{}\s]*$/.test(text.replace(/\
\(?:["\\\/bfnrt]|u[0-9a-fA-F]{4})/g,"@").replace(/"[^"\\\n\r]*"|true|false|null|-?\d+(?:\.\d*)?(?:[eE][+\-]?\d+)?/g,"]").replace(/(?:^|:|,)(?:\s*\[)+/g,""))){j=eval("("
+text+")");return typeof reviver=="function"?walk({"":j},""):j}throw new SyntaxError("JSON.parse")})}()
+
 
 //     [*] Including lib/index.js
 // Public object

@@ -231,18 +228,6 @@ utils.isSameOriginUrl = function(url_a, url_b) {
             url_b.split('/').slice(0,3).join('/'));
 };
 
-// <METEOR>
-// https://github.com/sockjs/sockjs-client/issues/79
-utils.isSameOriginScheme = function(url_a, url_b) {
-    if (!url_b) url_b = _window.location.href;
-
-    return (url_a.split(':')[0]
-                ===
-            url_b.split(':')[0]);
-};
-// </METEOR>
-
-
 utils.getParentDomain = function(url) {
     // ipv4 ip address
     if (/^[0-9.]*$/.test(url)) return url;

@@ -298,12 +283,6 @@ utils.log = function() {
     }
 };
 
-utils.debug = function() {
-    if (_window.console && console.debug && console.debug.apply) {
-        console.debug.apply(console, arguments);
-    }
-};
-
 utils.bind = function(fun, that) {
     if (fun.bind) {
         return fun.bind(that);

@@ -318,29 +297,8 @@ utils.flatUrl = function(url) {
     return url.indexOf('?') === -1 && url.indexOf('#') === -1;
 };
 
-// `relativeTo` is an optional absolute URL. If provided, `url` will be
-// interpreted relative to `relativeTo`. Defaults to `document.location`.
-// <METEOR>
-utils.amendUrl = function(url, relativeTo) {
-    var baseUrl;
-    if (relativeTo === undefined) {
-      baseUrl = _document.location;
-    } else {
-      var protocolMatch = /^([a-z0-9.+-]+:)/i.exec(relativeTo);
-      if (protocolMatch) {
-        var protocol = protocolMatch[0].toLowerCase();
-        var rest = relativeTo.substring(protocol.length);
-        var hostMatch = /[a-z0-9\.-]+(:[0-9]+)?/.exec(rest);
-        if (hostMatch)
-          var host = hostMatch[0];
-      }
-      if (! protocol || ! host)
-        throw new Error("relativeTo must be an absolute url");
-      baseUrl = {
-        protocol: protocol,
-        host: host
-      };
-    }
+utils.amendUrl = function(url) {
+    var dl = _document.location;
     if (!url) {
         throw new Error('Wrong url for SockJS');
     }

@@ -350,25 +308,14 @@ utils.amendUrl = function(url, relativeTo) {
 
     //  '//abc' --> 'http://abc'
     if (url.indexOf('//') === 0) {
-        url = baseUrl.protocol + url;
+        url = dl.protocol + url;
     }
-    // '/abc' --> 'http://localhost:1234/abc'
+    // '/abc' --> 'http://localhost:80/abc'
     if (url.indexOf('/') === 0) {
-        url = baseUrl.protocol + '//' + baseUrl.host + url;
+        url = dl.protocol + '//' + dl.host + url;
     }
-    // </METEOR>
     // strip trailing slashes
     url = url.replace(/[/]+$/,'');
-
-    // We have a full url here, with proto and host. For some browsers
-    // http://localhost:80/ is not in the same origin as http://localhost/
-       // Remove explicit :80 or :443 in such cases. See #74
-    var parts = url.split("/");
-    if ((parts[0] === "http:" && /:80$/.test(parts[2])) ||
-           (parts[0] === "https:" && /:443$/.test(parts[2]))) {
-               parts[2] = parts[2].replace(/:(80|443)$/, "");
-       }
-    url = parts.join("/");
     return url;
 };
 

@@ -987,11 +934,11 @@ utils.isXHRCorsCapable = function() {
  */
 
 var SockJS = function(url, dep_protocols_whitelist, options) {
-    if (!(this instanceof SockJS)) {
+    if (this === _window) {
         // makes `new` optional
         return new SockJS(url, dep_protocols_whitelist, options);
     }
-
+    
     var that = this, protocols_whitelist;
     that._options = {devel: false, debug: false, protocols_whitelist: [],
                      info: undefined, rtt: undefined};

@@ -1089,10 +1036,8 @@ SockJS.prototype._didClose = function(code, reason, force) {
     var that = this;
     if (that.readyState !== SockJS.CONNECTING &&
         that.readyState !== SockJS.OPEN &&
-        that.readyState !== SockJS.CLOSING) {
-        utils.debug('INVALID_STATE_ERR', that.readyState);
-        return;
-    }
+        that.readyState !== SockJS.CLOSING)
+            throw new Error('INVALID_STATE_ERR');
     if (that._ir) {
         that._ir.nuke();
         that._ir = null;


@@ -1238,23 +1183,8 @@ SockJS.prototype._applyInfo = function(info, rtt, protocols_whitelist) {
     that._options.rtt = rtt;
     that._options.rto = utils.countRTO(rtt);
     that._options.info.null_origin = !_document.domain;
-    // Servers can override base_url, eg to provide a randomized domain name and
-    // avoid browser per-domain connection limits.
-    if (info.base_url)
-      // <METEOR>
-      that._base_url = utils.amendUrl(info.base_url, that._base_url);
-      // </METEOR>
     var probed = utils.probeProtocols();
     that._protocols = utils.detectProtocols(probed, protocols_whitelist, info);
-// <METEOR>
-// https://github.com/sockjs/sockjs-client/issues/79
-    // Hack to avoid XDR when using different protocols
-    // We're on IE trying to do cross-protocol. jsonp only.
-    if (!utils.isSameOriginScheme(that._base_url) &&
-        2 === utils.isXHRCorsCapable()) {
-        that._protocols = ['jsonp-polling'];
-    }
-// </METEOR>
 };
 //         [*] End of lib/sockjs.js
 

@@ -2003,17 +1933,7 @@ InfoReceiver.prototype = new EventEmitter(['finish']);
 InfoReceiver.prototype.doXhr = function(base_url, AjaxObject) {
     var that = this;
     var t0 = (new Date()).getTime();
-
-// <METEOR>
-  // https://github.com/sockjs/sockjs-client/pull/129
-  // var xo = new AjaxObject('GET', base_url + '/info');
-
-    var xo = new AjaxObject(
-      // add cachebusting parameter to url to work around a chrome bug:
-      // https://code.google.com/p/chromium/issues/detail?id=263981
-      // or misbehaving proxies.
-      'GET', base_url + '/info?cb=' + utils.random_string(10))
-// </METEOR>
+    var xo = new AjaxObject('GET', base_url + '/info');
 
     var tref = utils.delay(8000,
                            function(){xo.ontimeout();});
@@ -2090,15 +2010,7 @@ var createInfoReceiver = function(base_url) {
         // XHRLocalObject -> no_credentials=true
         return new InfoReceiver(base_url, utils.XHRLocalObject);
     case 2:

-// <METEOR>
-// https://github.com/sockjs/sockjs-client/issues/79
-        // XDR doesn't work across different schemes
-        // http://blogs.msdn.com/b/ieinternals/archive/2010/05/13/xdomainrequest-restrictions-limitations-and-workarounds.aspx
-        if (utils.isSameOriginScheme(base_url))
-            return new InfoReceiver(base_url, utils.XDRObject);
-        else
-            return new InfoReceiverFake();
-// </METEOR>
+        return new InfoReceiver(base_url, utils.XDRObject);
     case 3:
         // Opera
         return new InfoReceiverIframe(base_url);
@@ -2464,3 +2376,4 @@ if (typeof define === 'function' && define.amd) {
 //     [*] End of lib/index.js
 
 // [*] End of lib/all.js
+

```

But It's important to keep in mind that the current PR adds the latest version so one needs to have a look at the [changelog](https://github.com/sockjs/sockjs-client/blob/master/Changelog.md).

filipenevola commented:
@harryadel about this

> using the NPM module would be a lot more elegant but I couldn't get it working as due to the reasons illustrated above. If you can get it working, I'd be happy to close this PR.

What were the reasons? Sorry I didn't find them.

harryadel commented:
@filipenevola The tests were failing and no matter how we tried importing/exporting the NPM package it didn't work out. Maybe you can give it a try.

> @StorytellerCZ no matter how `sockjs-client` is imported/exported, it always resulted in an error in one of the self tests which you can run locally using:
> 
> ```
> ./meteor self-test "add packages client archs"
> ```
> 
> Copying the contents of the CDN `https://cdn.jsdelivr.net/npm/sockjs-client@1/dist/sockjs.min.js` into `sockjs.js` seems to fix the problem. What do you think?

filipenevola commented:
@harryadel did you add the Meteor customizations to this version as well? If not, why they are not necessary anymore?

harryadel commented:
@filipenevola Re-adding the Meteor changes wasn't part of the plan because the tests passed if they didn't then I'd have gone back and added the needed parts accordingly — There was no reason to over-invest effort beforehand without real need. 

Keep in mind the diff showcased above entails the 0.3.4 version, since then the library has undergone some serious changes to its architecture making it no longer a simple single file. And to implement these changes back would be a little bit weird as we don't know what pushed the sockjs team to make such changes and even how these changes affect the Meteor setup! 

Actually writing this down just now made me realize that the problem might not be due to the importing but rather the huge version leap. I'mma start going up one by one to see if it fixes it. A not so recent  NPM import is much better than including the latest version of the file.

harryadel commented:
Updating the package version one by one didn't help, the immediate next available package version 1.0.0-beta.4 led to the same error 

```
 => Failure Reason: "match-timeout" at "tools/tests/package-tests.js:369"
  => Pattern: "all clients 0"                 
  => Last 100 lines:                          
  1| [[[[[ /tmp/mt-1oi6jl5.dwzji/home/myapp ]]]]]
  1|                                          
  1| => Started proxy.                        
  1| => Started your app.                     
  1|                                          
  1| => App running at: http://localhost:3000/
```

Here're the different ways we tried importing sockjs-client:

`import '.npm/package/node_modules/sockjs-client/dist/sockjs.min.js';`
`import { SockJS } from "sockjs-client"`
`import SockJS from "sockjs-client";`

I thought the culprit might by the legacy import line
`api.addFiles("sockjs-0.3.4.js", "legacy");`
but actually omitting this line with the 0.3.4 file doesn't stop the tests from running smoothly. :shrug:

harryadel commented:
![Screenshot_2021-07-16_09-02-44](https://user-images.githubusercontent.com/6644545/125906208-42ec0be7-bd35-4e02-9053-3c864331a3c1.png)
